Compilation error because of missing CUB_NS_QUALIFIER

The recent build of k2 fails because of this error:
```
/opt/cuda/include/cub/util_namespace.cuh:46:2: error:
#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
```
Reference in the CUB code: https://github.com/NVIDIA/cub/blob/94a50bf20cc01f44863a524ba36e089fd80f342e/cub/util_namespace.cuh#L43-L47

System is Arch Linux with latest libraries, with CUDA 11.5 and CUB 1.13.1.

<details><summary>Here is the full output:</summary><div>

At the most recent commit  e8c589a:
```bash
export K2_IS_STABLE=1   # To obtain a version number compatible with PEP 440
python setup.py build
```

```
running build
running build_py
creating build
creating build/lib.linux-x86_64-3.9
creating build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/symbol_table.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/nbest.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/__init__.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/autograd_utils.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/fsa_algo.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/ops.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/autograd.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/fsa_properties.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/dense_fsa_vec.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/utils.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/ctc_loss.py -> build/lib.linux-x86_64-3.9/k2
copying k2/python/k2/fsa.py -> build/lib.linux-x86_64-3.9/k2
creating build/lib.linux-x86_64-3.9/k2/ragged
copying k2/python/k2/ragged/__init__.py -> build/lib.linux-x86_64-3.9/k2/ragged
creating build/lib.linux-x86_64-3.9/k2/sparse
copying k2/python/k2/sparse/autograd.py -> build/lib.linux-x86_64-3.9/k2/sparse
copying k2/python/k2/sparse/__init__.py -> build/lib.linux-x86_64-3.9/k2/sparse
creating build/lib.linux-x86_64-3.9/k2/version
copying k2/python/k2/version/__init__.py -> build/lib.linux-x86_64-3.9/k2/version
copying k2/python/k2/version/version.py -> build/lib.linux-x86_64-3.9/k2/version
copying k2/python/k2/version/__main__.py -> build/lib.linux-x86_64-3.9/k2/version
running build_ext
Setting PYTHON_EXECUTABLE to /usr/bin/python
build command is:

                cd build/temp.linux-x86_64-3.9

                cmake -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=/usr/bin/python /xxx/python-k2-git

                cat k2/csrc/version.h

                make  _k2
            
-- Enabled languages: CXX;CUDA
-- The CXX compiler identification is GNU 11.1.0
-- The CUDA compiler identification is NVIDIA 11.5.50
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /opt/cuda/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- K2_OS: Arch Linux
-- Found Git: /usr/bin/git (found version "2.33.1") 
-- Looking for C++ include cxxabi.h
-- Looking for C++ include cxxabi.h - found
-- Looking for C++ include execinfo.h
-- Looking for C++ include execinfo.h - found
-- Performing Test K2_COMPILER_SUPPORTS_CXX14
-- Performing Test K2_COMPILER_SUPPORTS_CXX14 - Success
-- C++ Standard version: 14
-- Automatic GPU detection failed. Building for common architectures.
-- Autodetected CUDA architecture(s): 3.5;5.0;5.2;6.0;6.1;7.0;7.5;8.0;8.6;8.6+PTX
-- K2_COMPUTE_ARCH_FLAGS: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_52,code=sm_52;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_86,code=compute_86
-- K2_COMPUTE_ARCH_CANDIDATES 35;50;60;61;70;75;80;86
-- Adding arch 35
-- Adding arch 50
-- Adding arch 60
-- Adding arch 61
-- Adding arch 70
-- Adding arch 75
-- Adding arch 80
-- Adding arch 86
-- K2_COMPUTE_ARCHS: 35;50;60;61;70;75;80;86
-- Could NOT find Valgrind (missing: Valgrind_INCLUDE_DIR Valgrind_EXECUTABLE) 
-- Downloading pybind11
-- pybind11 is downloaded to /xxx/python-k2-git/build/temp.linux-x86_64-3.9/_deps/pybind11-src
-- pybind11 v2.6.0 
-- Found PythonInterp: /usr/bin/python (found version "3.9.7") 
-- Found PythonLibs: /usr/lib/libpython3.9.so
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Python executable: /usr/bin/python
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Caffe2: Found gflags with new-style gflags target.
-- Caffe2: Found glog with new-style glog target.
-- Caffe2: Found protobuf with new-style protobuf targets.
-- Caffe2: Protobuf version 3.17.3
-- Found CUDA: /opt/cuda (found version "11.5") 
-- Caffe2: CUDA detected: 11.5
-- Caffe2: CUDA nvcc is: /opt/cuda/bin/nvcc
-- Caffe2: CUDA toolkit directory: /opt/cuda
-- Caffe2: Header version is: 11.5
-- Found CUDNN: /usr/lib/libcudnn.so  
-- Found cuDNN: v8.2.4  (include: /usr/lib/python3.9/site-packages/torch/include, library: /usr/lib/libcudnn.so)
-- /opt/cuda/lib64/libnvrtc.so shorthash is 6dcccaf3
-- Automatic GPU detection failed. Building for common architectures.
-- Autodetected CUDA architecture(s): 3.5;5.0;5.2;6.0;6.1;7.0;7.5;8.0;8.6;8.6+PTX
-- Added CUDA NVCC flags for: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_52,code=sm_52;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_86,code=compute_86
CMake Warning at /usr/lib64/cmake/Torch/TorchConfig.cmake:22 (message):
  static library kineto_LIBRARY-NOTFOUND not found.
Call Stack (most recent call first):
  /usr/lib64/cmake/Torch/TorchConfig.cmake:127 (append_torchlib_if_found)
  cmake/torch.cmake:11 (find_package)
  CMakeLists.txt:256 (include)


-- Found Torch: /usr/lib/python3.9/site-packages/torch/lib/libtorch.so  
-- PyTorch version: 1.10.0
-- PyTorch cuda version: 11.5
-- Downloading moderngpu
-- moderngpu is downloaded to /xxx/python-k2-git/build/temp.linux-x86_64-3.9/_deps/moderngpu-src
-- Downloading googletest
-- googletest is downloaded to /xxx/python-k2-git/build/temp.linux-x86_64-3.9/_deps/googletest-src
-- googletest's binary dir is /xxx/python-k2-git/build/temp.linux-x86_64-3.9/_deps/googletest-build
CMake Deprecation Warning at build/temp.linux-x86_64-3.9/_deps/googletest-src/CMakeLists.txt:4 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Deprecation Warning at build/temp.linux-x86_64-3.9/_deps/googletest-src/googlemock/CMakeLists.txt:45 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


CMake Deprecation Warning at build/temp.linux-x86_64-3.9/_deps/googletest-src/googletest/CMakeLists.txt:56 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- CMAKE_CUDA_FLAGS:  --expt-extended-lambda -gencode arch=compute_35,code=sm_35 --expt-extended-lambda -gencode arch=compute_50,code=sm_50 --expt-extended-lambda -gencode arch=compute_60,code=sm_60 --expt-extended-lambda -gencode arch=compute_61,code=sm_61 --expt-extended-lambda -gencode arch=compute_70,code=sm_70 --expt-extended-lambda -gencode arch=compute_75,code=sm_75 --expt-extended-lambda -gencode arch=compute_80,code=sm_80 --expt-extended-lambda -gencode arch=compute_86,code=sm_86 -D_GLIBCXX_USE_CXX11_ABI=1 --compiler-options -Wall --compiler-options -Wno-unknown-pragmas --compiler-options -Wno-strict-overflow
-- CMAKE_CXX_FLAGS: -march=native -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2,-D_GLIBCXX_ASSERTIONS         -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-strict-overflow
-- Generated /xxx/python-k2-git/build/temp.linux-x86_64-3.9/k2/csrc/version.h
-- PYTHON_LIBRARIES: /usr/lib/libpython3.9.so
-- Configuring done
-- Generating done
-- Build files have been written to: /xxx/python-k2-git/build/temp.linux-x86_64-3.9
/**
 * @brief
 * This file contains information for building k2, which is helpful
 * for describing the environment in which k2 is used when creating
 * an issue on GitHub.
 *
 * @copyright
 * Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
 *
 * @copyright
 * See LICENSE for clarification regarding multiple authors
 *
 * @caution
 * This file is auto-generated by CMake. Do NOT edit!
 * You can find it at build_dir/k2/csrc/version.h
 */

#ifndef K2_CSRC_VERSION_H_
#define K2_CSRC_VERSION_H_

namespace k2 {

// Version of k2 in "major.minor.patch" format
static constexpr const char *kVersion = "1.10";

// The commit used to build k2
static constexpr const char *kGitSha1 = "e8c589a47e8acb19c9c08df5ceaba7d19d78238e";

// Date of the commit used to build k2
static constexpr const char *kGitDate = "Thu Nov 4 15:10:40 2021";

// Version of CUDA used to build k2.
// Its format is "major.minor", e.g., 10.1
static constexpr const char *kCudaVersion = "11.5";

// cuDNN version, e.g., 8.0.2
static constexpr const char *kCudnnVersion = "8.2.4";

// clang-format off
// Version of Python used to build k2 Python bindings.
static constexpr const char *kPythonVersion = "3.9";

// clang-format on

// CMake build type, e.g., Release or Debug.
static constexpr const char *kBuildType = "Release";

// The operating system that is used to build k2, e.g., Ubuntu 16.04 LTS
static constexpr const char *kOS = "Arch Linux";

// e.g., 3.18.0
static constexpr const char *kCMakeVersion = "3.21.4";

// Version of the compiler, e.g., 5.4.0
static constexpr const char *kGCCVersion = "11.1.0";

// CUDA flags used to compile k2
static constexpr const char *kCMakeCudaFlags = " --expt-extended-lambda -gencode arch=compute_35,code=sm_35 --expt-extended-lambda -gencode arch=compute_50,code=sm_50 --expt-extended-lambda -gencode arch=compute_60,code=sm_60 --expt-extended-lambda -gencode arch=compute_61,code=sm_61 --expt-extended-lambda -gencode arch=compute_70,code=sm_70 --expt-extended-lambda -gencode arch=compute_75,code=sm_75 --expt-extended-lambda -gencode arch=compute_80,code=sm_80 --expt-extended-lambda -gencode arch=compute_86,code=sm_86 -D_GLIBCXX_USE_CXX11_ABI=1 --compiler-options -Wall --compiler-options -Wno-unknown-pragmas --compiler-options -Wno-strict-overflow";

// CXX flags used to compile k2
static constexpr const char *kCMakeCxxFlags = "-march=native -O2 -pipe -fno-plt -fexceptions         -Wp,-D_FORTIFY_SOURCE=2,-D_GLIBCXX_ASSERTIONS         -Wformat -Werror=format-security         -fstack-clash-protection -fcf-protection -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-strict-overflow";

// Which PyTorch version k2 is using, e.g., 1.6.0+cu101
static constexpr const char *kTorchVersion = "1.10.0";

// Which CUDA version PyTorch is using, e.g., 10.1
static constexpr const char *kTorchCudaVersion = "11.5";

#ifndef K2_WITH_CUDA
#define K2_WITH_CUDA
#endif

#ifdef K2_WITH_CUDA
  static constexpr bool kWithCuda = true;
#else
  static constexpr bool kWithCuda = false;
#endif

// Indicate whether NVTX is enabled or not
#ifndef K2_ENABLE_NVTX
#define K2_ENABLE_NVTX
#endif

#ifdef K2_ENABLE_NVTX
static constexpr bool kEnableNvtx = true;
#else
static constexpr bool kEnableNvtx = false;
#endif
}  // namespace k2

#endif  //  K2_CSRC_VERSION_H_
[  0%] Building CUDA object k2/csrc/CMakeFiles/k2_log.dir/log.cu.o
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
[  3%] Linking CUDA shared library ../../lib/libk2_log.so
[  3%] Built target k2_log
[  6%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/arcsort.cc.o
[  9%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/determinize.cc.o
[  9%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/fsa.cc.o
[  9%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/aux_labels.cc.o
[ 12%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/determinize_pruned.cc.o
[  9%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/connect.cc.o
[ 15%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/fsa_equivalent.cc.o
[ 15%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/fsa_renderer.cc.o
[ 18%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/fsa_util.cc.o
[ 18%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/intersect.cc.o
[ 21%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/properties.cc.o
[ 21%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/rmepsilon_pruned.cc.o
[ 24%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/topsort.cc.o
[ 24%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/util.cc.o
[ 27%] Building CXX object k2/csrc/host/CMakeFiles/fsa.dir/weights.cc.o
[ 27%] Linking CXX shared library ../../../lib/libk2fsa.so
[ 27%] Built target fsa
[ 30%] Building CUDA object k2/csrc/CMakeFiles/context.dir/array_ops.cu.o
[ 33%] Building CUDA object k2/csrc/CMakeFiles/context.dir/context.cu.o
[ 30%] Building CUDA object k2/csrc/CMakeFiles/context.dir/algorithms.cu.o
[ 33%] Building CUDA object k2/csrc/CMakeFiles/context.dir/connect.cu.o
[ 36%] Building CUDA object k2/csrc/CMakeFiles/context.dir/fsa.cu.o
[ 36%] Building CUDA object k2/csrc/CMakeFiles/context.dir/dtype.cu.o
[ 36%] Building CUDA object k2/csrc/CMakeFiles/context.dir/fsa_algo.cu.o
[ 39%] Building CUDA object k2/csrc/CMakeFiles/context.dir/fsa_utils.cu.o
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/array_ops.h:27,
                 from /xxx/python-k2-git/k2/csrc/fsa_algo.cu:28:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/array_ops.h:27,
                 from /xxx/python-k2-git/k2/csrc/fsa.cu:12:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/algorithms.h:25,
                 from /xxx/python-k2-git/k2/csrc/algorithms.cu:22:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/array_ops.h:27,
                 from /xxx/python-k2-git/k2/csrc/connect.cu:22:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/array_ops.cu:22:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
In file included from /opt/cuda/include/cub/util_arch.cuh:37,
                 from /opt/cuda/include/cub/config.cuh:35,
                 from /opt/cuda/include/cub/cub.cuh:37,
                 from /xxx/python-k2-git/k2/csrc/cub.h:37,
                 from /xxx/python-k2-git/k2/csrc/array_inl.h:33,
                 from /xxx/python-k2-git/k2/csrc/array.h:969,
                 from /xxx/python-k2-git/k2/csrc/fsa_utils.cu:32:
/opt/cuda/include/cub/util_namespace.cuh:46:2: error: #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
   46 | #error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.
      |  ^~~~~
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:76: k2/csrc/CMakeFiles/context.dir/algorithms.cu.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:90: k2/csrc/CMakeFiles/context.dir/array_ops.cu.o] Error 1
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:146: k2/csrc/CMakeFiles/context.dir/fsa.cu.o] Error 1
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:104: k2/csrc/CMakeFiles/context.dir/connect.cu.o] Error 1
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:174: k2/csrc/CMakeFiles/context.dir/fsa_utils.cu.o] Error 1
make[3]: *** [k2/csrc/CMakeFiles/context.dir/build.make:160: k2/csrc/CMakeFiles/context.dir/fsa_algo.cu.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:810: k2/csrc/CMakeFiles/context.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2015: k2/python/csrc/CMakeFiles/_k2.dir/rule] Error 2
make: *** [Makefile:823: _k2] Error 2
Traceback (most recent call last):
  File "/xxx/python-k2-git/setup.py", line 198, in <module>
    setuptools.setup(
  File "/usr/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.9/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.9/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.9/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/usr/lib/python3.9/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.9/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 79, in run
    _build_ext.run(self)
  File "/usr/lib/python3.9/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
    _build_ext.build_ext.run(self)
  File "/usr/lib/python3.9/distutils/command/build_ext.py", line 340, in run
    self.build_extensions()
  File "/usr/lib/python3.9/site-packages/Cython/Distutils/old_build_ext.py", line 195, in build_extensions
    _build_ext.build_ext.build_extensions(self)
  File "/usr/lib/python3.9/distutils/command/build_ext.py", line 449, in build_extensions
    self._build_extensions_serial()
  File "/usr/lib/python3.9/distutils/command/build_ext.py", line 474, in _build_extensions_serial
    self.build_extension(ext)
  File "/xxx/python-k2-git/setup.py", line 146, in build_extension
    raise Exception('Failed to build k2')
Exception: Failed to build k2
```

</div></details>


Thanks!

--- a/cmake/cub.cmake
+++ b/cmake/cub.cmake
@@ -20,8 +20,9 @@ function(download_cub)
 
   include(FetchContent)
 
-  set(cub_URL  "https://github.com/NVlabs/cub/archive/1.10.0.tar.gz")
-  set(cub_HASH "SHA256=8531e09f909aa021125cffa70a250761dfc247f960d7a1a12f65e6651ffb6477")
+  set(cub_URL  "https://github.com/NVlabs/cub/archive/1.15.0.tar.gz")
+  set(cub_HASH "SHA256=1781ee5eb7f00acfee5bff88e3acfc67378f6b3c24281335e18ae19e1f2ff685")
+
 
   FetchContent_Declare(cub
     URL               ${cub_URL}

--- a/k2/csrc/cub.h
+++ b/k2/csrc/cub.h
@@ -30,8 +30,33 @@
 // that k2 and PyTorch use a different copy
 // of CUB.
 
+#ifdef CUB_NS_PREFIX
+#undef CUB_NS_PREFIX
+#endif
+
+#ifdef CUB_NS_POSTFIX
+#undef CUB_NS_POSTFIX
+#endif
+
+#ifdef CUB_NS_QUALIFIER
+#undef CUB_NS_QUALIFIER
+#endif
+
+// see
+// https://github.com/NVIDIA/cub/commit/6631c72630f10e370d93814a59146b12f7620d85
+// The above commit replaced "thrust" with "THRUST_NS_QUALIFIER"
+#ifndef THRUST_NS_QUALIFIER
+#define THRUST_NS_QUALIFIER thrust
+#endif
+
 #define CUB_NS_PREFIX namespace k2 {
 #define CUB_NS_POSTFIX }
+
+// See
+// https://github.com/NVIDIA/cub/commit/6631c72630f10e370d93814a59146b12f7620d85
+// and
+// https://github.com/NVIDIA/cub/pull/350
+#define CUB_NS_QUALIFIER ::k2::cub
 
 #ifdef K2_WITH_CUDA
 #include "cub/cub.cuh"  // NOLINT
@@ -39,5 +64,6 @@
 
 #undef CUB_NS_PREFIX
 #undef CUB_NS_POSTFIX
+#undef CUB_NS_QUALIFIER
 
 #endif  // K2_CSRC_CUB_H_

